academy/4-my-own-chain/server-side.md

```diff
@@ -0,0 +1,1046 @@
+        process.stdout.cursorTo(0)
+        // Get the block
+        const block: Block = await client.getBlock(processing)
+        process.stdout.write(`Handling block: ${processing} with ${block.txs.length} txs`)
+        // Function yet to be declared
+        await handleBlock(block)
+        db.status.block.height = processing
+    }
+    await saveDb()
+    timer = setTimeout(poll, pollIntervalMs)
+}
+```
+
+This needs a new import:
+
+```typescript [https://github.com/cosmos/academy-checkers-ui/blob/b030b2f/src/server/indexer.ts#L3]
+import { Block } from "@cosmjs/stargate"
+```
+
+The indexer:
+
+* Declares a new function `handleBlock`. Create one and put `console.log(block)` inside to explore what this object is and consider what actions you would take with it.
+* Saves the `db` after a poll, so _you_ can watch it in real time.
+* Uses `process.stdout.write` and `process.stdout.cursorTo(0)` so that the repetitive logging happens on a single line.
+
+Observe the relevant content in `handleBlock`. It must:
+
+1. Extract the events from transactions.
+2. Extract the events from `EndBlock`.
+
+If you examine `block.txs` directly you find transactions as they were posted. However, this does not reveal any execution results, such as if a transaction executed as expected or what game ID it used for the new game. To get this extra information:
+
+1. Calculate `txHash` from the transaction.
+2. Call `await client.getTx(txHash)`, which returns an `IndexedTx`.
+
+The `handleBlock` function can be:
+
+```typescript [https://github.com/cosmos/academy-checkers-ui/blob/b030b2f/src/server/indexer.ts#L81-L92]
+const handleBlock = async (block: Block) => {
+    if (0 < block.txs.length) console.log("")
+    let txIndex = 0
+    while (txIndex < block.txs.length) {
+        const txHash: string = toHex(sha256(block.txs[txIndex])).toUpperCase()
+        const indexed: IndexedTx | null = await client.getTx(txHash)
+        if (!indexed) throw new Error(`Could not find indexed tx: ${txHash}`)
+        // Function yet to be declared
+        await handleTx(indexed)
+        txIndex++
+    }
+    // TODO handle EndBlock
+}
+```
+
+This needs new imports:
+
+```typescript [https://github.com/cosmos/academy-checkers-ui/blob/b030b2f/src/server/indexer.ts#L1-L3]
+import { sha256 } from "@cosmjs/crypto"
+import { toHex } from "@cosmjs/encoding"
+import { Block, IndexedTx } from "@cosmjs/stargate"
+```
+
+* `while() {}` simplifies the syntax of `await`ing multiple times.
+* The hash is calculated this way as per [here](https://github.com/cosmos/cosmjs/blob/902f21b/packages%2Fstargate%2Fsrc%2Fstargateclient.ts#L74).
+* `console.log("")` puts a new line (`poll` does a `process.stdout.write` which adds no line).
+* The `handleBlock` function uses a new function, `handleTx`. Create one and put `console.log(indexed)` inside to explore what this object is and consider what actions you can take with it.
+* The `EndBlock` part has not yet been incorporated.
```

Comment:
Is it helpful to mention something that has not been done? If not, I would delete. 

If so, maybe this line should be extended to direct attention to the relevant part of the text, ie:

"Incorporating the `EndBlock` part is explained in **Prepare for EndBlock**."

Reply:
Yes, I added it only for the reader to keep a mental tab on what comes later. It is my way of helping the reader not get too lost on such a long page. What I want to avoid is someone going through the motions mindlessly, not seeing further than 1 paragraph and in the end not understanding what just happened.
Then, when the reader reaches the end block part, their brain goes "ah yes yes, that was remaining to be done". And the association with the rest will be hopefully that it is on the same level as `handleTx`, even if it appears way down:
- handle block
    - handle tx
        - handle events
            - handle event
    - handle end block

Reply:
In other places we have made this foreshadowing more explicit, where the top of the page describes what we will do.